### PR TITLE
Feature/adl 36

### PR DESCRIPTION
**Pull request by S3784866**
Footer bar
- Displays CarID and UTC Time from AV Logger.
- Created basic Heartbeat message relay and displays alongside connection.
- Raw AV Data logger messages displayed in Console box with auto scroll function.

### CODE REVIEW

File: Dashboard/Utils/DependencyInjection.cs [hunk 1]
```diff
@@ -19,11 +19,13 @@ public static IServiceProvider ConfigureServices()
 
         services.AddSingleton<IDataStore, DataStore>();
         services.AddSingleton<IConnector, SerialConnector>();
+        services.AddSingleton<ISerialPort, SerialPortWrapper>();
 
         services.AddTransient<DataViewModel>();
         services.AddTransient<MainWindowViewModel>();
         services.AddTransient<ISerialPort, SerialPortWrapper>();
```
Comment (nrdag) [inline Dashboard/Utils/DependencyInjection.cs:25]: These are doubled up now, i think i put this as transient so we can potentially recreate this at runtime via the connection page, but not quite sure how that would work actually

File: Dashboard/ViewModels/FooterViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+{
+    public bool? HeartBeat => _dataStore.HeartBeat;
+    public event Action<bool>? HeartbeatStatusUpdated;
+    public event Action<string>? RawMessageUpdated;
+    public event Action<bool>? ConnectionUpdate;
+    private readonly IDataStore _dataStore;
+    public bool ConnectionStatus => _dataStore.RawData?.ConnectionStatus ?? false;
+    public string CarId => _dataStore.RawData?.CarId ?? "0";
+    public string RawMessage => _dataStore.RawData?.RawMessage ?? "";
+    public string UTcTime
```
Comment (nrdag) [inline Dashboard/ViewModels/FooterViewModel.cs]: Change var name to UTCTime plz

File: Dashboard/ViewModels/FooterViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        Console.WriteLine("Heart Beat change triggered");
+        OnPropertyChanged(nameof(HeartBeat));
+        HeartbeatStatusUpdated?.Invoke(isReceived);
+    }
+
+    public override void Dispose()
+    {
+        _dataStore.RawDataUpdated -= OnRawDataChanged;
+        _dataStore.HeartBeatUpdated -= OnHeartbeatStatusChanged;
+        // _dataStore.Dispose();
```
Comment (nrdag) [inline Dashboard/ViewModels/FooterViewModel.cs]: plz remove commented code

File: Dashboard/Models/DataStore.cs [hunk 4]
```diff
@@ -8,10 +8,14 @@ public class DataStore : IDataStore, IDisposable
     public event EventHandler? GpsDataUpdated;
     public event EventHandler? AvDataUpdated;
     public event EventHandler? ResDataUpdated;
+    public event EventHandler? RawDataUpdated;
```
Comment (nrdag) [inline Dashboard/Models/DataStore.cs:11]: I am cautious about adding this, I am unsure if we want to just straight output everything via events, it may be nicer if we do this via a log file, and either have some log event that gets consumed by our view model or we tail a log file for the console?

File: Dashboard/ViewModels/FooterViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        // Convert the UTC value from data logger to proper UTC.
+        get
+        {
+            var utcTimeString = _dataStore.RawData?.UTCTime;
+            if (string.IsNullOrEmpty(utcTimeString))
+            {
+                return "Invalid time";
+            }
+            try
```
Comment (nrdag) [inline Dashboard/ViewModels/FooterViewModel.cs]: Could we do this conversion at the serial layer? So everything is consuming proper formatted times? (maybe convert to a datetime object and pass that around)

File: Dashboard/Connectors/Serial/SerialConnector.cs [hunk 6]
```diff
@@ -10,6 +11,10 @@ public class SerialConnector(ISerialPort comPort) : IConnector
     public event EventHandler<GpsData>? GpsDataUpdated;
     public event EventHandler<AvData>? AvDataUpdated;
     public event EventHandler<ResData>? ResDataUpdated;
+    public event EventHandler<RawData>? RawDataUpdated;
+    public event EventHandler<bool>? HeartBeatUpdated;
+    private Timer? _heartbeatTimer;
+    private const string HeartBeatMessage = "HeartBeat";
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialConnector.cs]: According to the spec, this should just be `CON?`

Comment (nrdag) [reply Dashboard/Connectors/Serial/SerialConnector.cs]: And the logger should return `OK` to signal that a device is connected to the logger

Comment (nrdag) [reply Dashboard/Connectors/Serial/SerialConnector.cs]: are we meant to be heart beat checking PC -> AV logger, or AV logger -> car?

File: Dashboard/Connectors/Serial/SerialConnector.cs [hunk 7]
```diff
@@ -19,6 +24,12 @@ public void Start()
         // Set up the event handler for when data is received
         comPort.DataReceived += OnDataReceived;
 
+        // Setting up timer for heatbeat messages
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialConnector.cs]: Do we need to be sending a heart beat every second, can we just check that the last received message was within some timeframe, and if not we send it?

File: Dashboard/Connectors/Serial/SerialConnector.cs [hunk 8]
```diff
@@ -38,18 +49,32 @@ public void Stop()
...
 
+    public bool SkipedFirst = false;
+
     private void ParseMessage(string message)
     {
         // All messages start with the format "#ID=<ID>|UTC=<TIME>|<MSG>", so we split the message by the '|', and remove the first 2 elements
         // First we validate that we have the correct message format
         if (!message.StartsWith("#ID=") || !message.Contains("|UTC="))
             throw new InvalidOperationException("Invalid message format received", new Exception(message));
+        // Very dirty method right now, Need to ask Nick how to better handle the first message being out of format
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialConnector.cs]: I fixed this stuff in ADL-45, this is a reminder for me to ensure this is fixed post merge (whichever ticket goes first)

File: Dashboard/Connectors/Serial/SerialConnector.cs [hunk 9]
```diff
@@ -161,4 +198,19 @@ private bool ParseBool(string value)
         return value.StartsWith('#') ? _random.Next() % 2 == 0 : value == "1";
     }
 
+    private void SendHeartbeat(object? state, ElapsedEventArgs e)
+    {
+        if (comPort.IsConnected)
+        {
+
+            Console.WriteLine($"Sending heartbeat: {HeartBeatMessage}");
+            comPort.Write(HeartBeatMessage);
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialConnector.cs]: For now we could have some fake implementation that returns the `OK` message when we do a connection test?

Comment (nrdag) [reply Dashboard/Connectors/Serial/SerialConnector.cs]: so we can implement the handling of that now rather than later

File: Dashboard/Connectors/Serial/SerialPortWrapper.cs [hunk 10]
```diff
@@ -65,4 +65,21 @@ public bool IsConnected
...
     }
+
+    public void Write(string? data)
+    {
+        if (_serialPort.IsOpen)
+        {
+            _serialPort.WriteTimeout = 50;
+            try
+            {
+                _serialPort.Write(data);
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialPortWrapper.cs:87]: q: does this block the main thread? do we need to consider opening a thread that writes to the port also?

File: Dashboard/ViewModels/FooterViewModel.cs (hunk 5 above)
Comment (nrdag) [reply Dashboard/ViewModels/FooterViewModel.cs]: additional q: do we even need to pull this from the av logger? can we just do this ourselves and get the current time in C#?

File: Dashboard/ViewModels/MainWindowViewModel.cs [hunk 11]
```diff
@@ -39,10 +42,15 @@ public partial class MainWindowViewModel : ViewModelBase
         /// </summary>
         public MainWindowViewModel()
         {
+
             _views = new Dictionary<string, (UserControl, ViewModelBase?)>();
             _serviceProvider = DependencyInjection.ConfigureServices();
             Items = new ObservableCollection<ListItemTemplate>(_templates);
             SelectedListItem = Items.First();
+            _footer = new FooterView
```
Comment (nrdag) [inline Dashboard/ViewModels/MainWindowViewModel.cs:45]: Does the footer change at runtime? Can we not just put the usercontrol directly in the MainWindowView.axaml XML?

File: Dashboard/Views/MainWindowView.axaml [hunk 12]
```diff
@@ -51,5 +51,8 @@
     <Grid Grid.Row="1">
         <ContentControl Name="MainContent" Content="{Binding CurrentPage}"/> 
     </Grid>
+        <Grid Grid.Row="2">
+            <ContentControl Name="FooterContent" Content="{Binding Footer}"/>
```
Comment (nrdag) [inline Dashboard/Views/MainWindowView.axaml:53]: can just be the footer directly here? <Footer>

**Review by nrdag — CHANGES_REQUESTED** [review]

File: Dashboard/ViewModels/MainWindowViewModel.cs (hunk 11 above)
Comment (S3784866) [reply Dashboard/ViewModels/MainWindowViewModel.cs:45]: The Footer could be directly a part of MainWindowView but it would come down to if we want to keep the Navigation system and Footer logic separate for maintaining. I opt'ed to keep them separate just to air on the side of being safe on that regard.

File: Dashboard/ViewModels/FooterViewModel.cs (hunk 5 above)
Comment (S3784866) [reply Dashboard/ViewModels/FooterViewModel.cs]: We could do this at the serial layer so other pages could use the times if needed.

As for not pulling it from the logger, I think it would depend on if we're making this in prep for replaying logs and if Nick O wants this UTC time to be the current or time of the original log.

File: Dashboard/ViewModels/FooterViewModel.cs (hunk 3 above)
Comment (S3784866) [reply Dashboard/ViewModels/FooterViewModel.cs]: Will do.
https://github.com/nickowen44/AV-Datalogger/pull/10/commits/dfa7244d705bf6043b7764f4c730fbe36a0eb0a2

File: Dashboard/ViewModels/FooterViewModel.cs (hunk 2 above)
Comment (S3784866) [reply Dashboard/ViewModels/FooterViewModel.cs]: Will do.
https://github.com/nickowen44/AV-Datalogger/pull/10/commits/dfa7244d705bf6043b7764f4c730fbe36a0eb0a2

File: Dashboard/Utils/DependencyInjection.cs (hunk 1 above)
Comment (S3784866) [reply Dashboard/Utils/DependencyInjection.cs:25]: I didn't even realized I had done this, I'll remove it.

File: Dashboard/Connectors/Serial/SerialPortWrapper.cs (hunk 10 above)
Comment (S3784866) [reply Dashboard/Connectors/Serial/SerialPortWrapper.cs:87]: Now that you've brought it up, I think it can block the main thread, I'll open a thread to handle this.

File: Dashboard/Connectors/Serial/SerialConnector.cs (hunk 7 above)
Comment (nrdag) [reply Dashboard/Connectors/Serial/SerialConnector.cs]: According to Nick O, we should consider sending the heartbeat only when we haven't received a message in some predetermined amount of time

File: Dashboard/ViewModels/MainWindowViewModel.cs (hunk 11 above)
Comment (nrdag) [reply Dashboard/ViewModels/MainWindowViewModel.cs:45]: I am thinking the view (or control) can remain separate, thats fine. my question was just can we do this statically via the XAML? so no need to dynamically create it in this file

Comment (S3784866) [reply Dashboard/ViewModels/MainWindowViewModel.cs:45]: I don't think so since the footer needs its own DataContext if we keep them separate from my understanding.

File: Dashboard/Models/DataStore.cs (hunk 4 above)
Comment (S3784866) [reply Dashboard/Models/DataStore.cs:11]: I'm also not too sure about this, I think it should be fine for right now and maybe we can loop back when we get the logging done?

Comment (nrdag) [reply Dashboard/Models/DataStore.cs:11]: That works for me, can you add to the AC for the logging ticket to do this? Maybe might need to re-estimate the ticket?

File: Dashboard/ViewModels/FooterViewModel.cs (hunk 5 above)
Comment (nrdag) [reply Dashboard/ViewModels/FooterViewModel.cs]: Discussed in the meeting with nick yesterday, will use the time from the logger as it will allow more flexibility when we add replay from file

File: Dashboard/Connectors/Serial/SerialConnector.cs (hunk 6 above)
Comment (nrdag) [reply Dashboard/Connectors/Serial/SerialConnector.cs]: > are we meant to be heart beat checking PC -> AV logger, or AV logger -> car?

we should be checking pc -> logger (serial connection)

File: Dashboard/ViewModels/MainWindowViewModel.cs (hunk 11 above)
Comment (nrdag) [reply Dashboard/ViewModels/MainWindowViewModel.cs:45]: Yep, I just tried it myself, I see what you mean. All good :)

File: Dashboard/Connectors/Serial/SerialPortWrapper.cs [hunk 13]
```diff
@@ -76,4 +76,30 @@ public bool IsConnected
...
+
+                // If the AV Logger doesn't respond OK then connection is dead.
+                if (_serialPort.ReadLine() != "OK")
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialPortWrapper.cs:90]: I don't think there is a guarantee that the next message read will be "OK", we should maybe add this as a message type to parse.

File: Dashboard/Connectors/Serial/SerialConnector.cs [hunk 14]
```diff
@@ -23,6 +33,23 @@ public void Start()
         // Set up the event handler for when data is received
         comPort.DataReceived += OnDataReceived;
 
+        //Set up Heart Beat thread 
+        _heartbeatThread = new Thread(() =>
+        {
+            while (_heartBeatShouldRun)
+            {
+                var timeSinceLastMessage = DateTime.Now - _lastMessageReceived;
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialConnector.cs]: There is tracking done within SerialPortWrapper -> IsConnected for this type of logic already, you could poll that instead. It currently checks if it has received any message within X seconds and if the port is open

File: Dashboard/Connectors/Serial/SerialConnector.cs [hunk 15]
```diff
@@ -48,8 +80,7 @@ private void ParseMessage(string message)
         if (!MyRegex().IsMatch(message))
             throw new InvalidOperationException("Invalid message format received", new Exception(message));
 
-        // All messages start with the format "#ID=<ID>|UTC=<TIME>|<MSG>", so we split the message by the '|', and remove the first 2 elements
-        var split = message.Split('|')[2..];
+        var split = message.Substring(1).Split('|');
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialConnector.cs:79]: What does this change mean? Is it for parsing a potential "OK" message? Could consider adding a strict eq check before the regex match on L80

Comment (nrdag) [reply Dashboard/Connectors/Serial/SerialConnector.cs:79]: (same as change on the L94 and L98)

File: Dashboard/Connectors/Serial/SerialConnector.cs [hunk 16]
```diff
@@ -143,6 +175,42 @@ private void ParseResMessage(Dictionary<string, string> values)
         });
     }
 
+    private DateTime ParseUTCTime(string timeString)
+    {
+
+        string datePart = "";
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialConnector.cs:177]: Good job on this, I don't like that it is required tbh. I think it would be best if we could push back and ask for an ISO 8601 str, but i think too much effort

File: Dashboard/Connectors/Serial/SerialConnector.cs [hunk 17]
```diff
@@ -143,6 +175,42 @@ private void ParseResMessage(Dictionary<string, string> values)
...
+            datePart = timeString.Substring(1, 4) + "0" + timeString.Substring(5, 3); // Extracts "2024820"
+        }
+        else
+        {
+            datePart = timeString.Substring(1, 8);
+        }
+
+        string timePart = timeString.Substring(9, 8);
+
+        DateTime parsedDate = DateTime.ParseExact(datePart, "yyyyMMdd", CultureInfo.InvariantCulture);
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialConnector.cs]: is there reasoning behind these being separate, im guessing due to formatting issues

File: Dashboard/Connectors/Serial/SerialPortWrapper.cs (hunk 13 above)
Comment (nrdag) [reply Dashboard/Connectors/Serial/SerialPortWrapper.cs:90]: my meaning behind ^ is there is a potential case where there is another message currently in the buffer, given we run this multi threaded we can't guarantee the order of things

File: Dashboard/Models/RawData.cs [hunk 18]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Dashboard.Models;
+
+public record RawData
+{
+    public string CarId { get; init; }
```
Comment (nrdag) [inline Dashboard/Models/RawData.cs]: U can fix this warning by adjusting all (or just hte ones that warn) to look like
```suggestion
    public required string CarId { get; init; }
```

Comment (nrdag) [reply Dashboard/Models/RawData.cs]: just means when the constructor or init statement is exited, it has to have a value

File: Dashboard/Views/FooterView.axaml [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+        </Style>
+    </UserControl.Styles>
+    <Design.DataContext>
+        <!-- This only sets the DataContext for the previewer in an IDE,
+             to set the actual DataContext for runtime, set the DataContext property in code (look at App.axaml.cs) -->
+        <vm:FooterViewModel />
+    </Design.DataContext>
+    <Grid ColumnDefinitions="0.2*,0.2*, 0.6*" >
+        <Grid Grid.Column="0">
+            <StackPanel VerticalAlignment="Center">
```
Comment (nrdag) [inline Dashboard/Views/FooterView.axaml:39]: Can you play around with the padding / margins plz:
- On the left of the times can you match the same margin that is to the right of hte console box
- Between the radios can you add a bit of padding

**Review by nrdag — CHANGES_REQUESTED** [review]

File: Dashboard/Connectors/Serial/SerialPortWrapper.cs (hunk 13 above)
Comment (S3784866) [reply Dashboard/Connectors/Serial/SerialPortWrapper.cs:90]: Ya, I wasn't too sure how we were going to handle this. I don't think we're able to test it since the sim doesn't respond to commands

File: Dashboard/Connectors/Serial/SerialConnector.cs (hunk 14 above)
Comment (S3784866) [reply Dashboard/Connectors/Serial/SerialConnector.cs]: I can swap the check to use IsConnected, I don't know why I didn't use that in the first place considering I was using IsConnected before the threading.

File: Dashboard/Connectors/Serial/SerialConnector.cs (hunk 15 above)
Comment (S3784866) [reply Dashboard/Connectors/Serial/SerialConnector.cs:79]: 1. It's to grab the ID and UTC time while trimming the first # off the ID.
2. I believe that was something I put in place when I brought up the buffer issue, you had asked to change the check to check for < 2 rather than != 2 as a stopgap. I can change it back to != 2.
3.  I changed it from using the Add function so it handles double up key-value pairs since we weren't checking for double ups. Currently I have it using the latest value of the key-value pair.

File: Dashboard/Connectors/Serial/SerialConnector.cs (hunk 16 above)
Comment (S3784866) [reply Dashboard/Connectors/Serial/SerialConnector.cs:181]: It would be nice to have it already be in a standard format. We can bring it up to Nick O next meeting if we have time.

File: Dashboard/Connectors/Serial/SerialConnector.cs (hunk 17 above)
Comment (S3784866) [reply Dashboard/Connectors/Serial/SerialConnector.cs]: Ya, I was having format issues, although I think I know how to shorten this to just use one variable.

File: Dashboard/Models/RawData.cs (hunk 18 above)
Comment (S3784866) [reply Dashboard/Models/RawData.cs]: Can do.

File: Dashboard/Views/FooterView.axaml (hunk 19 above)
Comment (S3784866) [reply Dashboard/Views/FooterView.axaml:39]: Can do.

File: Dashboard/Connectors/Serial/SerialConnector.cs (hunk 16 above)
Comment (S3784866) [reply Dashboard/Connectors/Serial/SerialConnector.cs:177]: It would be good to have it already in a standard format, We could bring it up to Nick O but given the timeline we currently have, its not a big issue.

File: Dashboard/Views/FooterView.axaml (hunk 19 above)
Comment (S3784866) [reply Dashboard/Views/FooterView.axaml:39]: https://github.com/nickowen44/AV-Datalogger/pull/10/commits/af9d6f189027cdae3bff3d19679e23588c2190d4

File: Dashboard/Models/RawData.cs (hunk 18 above)
Comment (S3784866) [reply Dashboard/Models/RawData.cs]: https://github.com/nickowen44/AV-Datalogger/pull/10/commits/af9d6f189027cdae3bff3d19679e23588c2190d4

File: Dashboard/Connectors/Serial/SerialConnector.cs (hunk 17 above)
Comment (S3784866) [reply Dashboard/Connectors/Serial/SerialConnector.cs]: https://github.com/nickowen44/AV-Datalogger/pull/10/commits/af9d6f189027cdae3bff3d19679e23588c2190d4

File: Dashboard/Connectors/Serial/SerialConnector.cs (hunk 14 above)
Comment (S3784866) [reply Dashboard/Connectors/Serial/SerialConnector.cs]: https://github.com/nickowen44/AV-Datalogger/pull/10/commits/af9d6f189027cdae3bff3d19679e23588c2190d4

**Comment by nrdag** [comment]
LGTM
